File: Dev-AI-PM/backend/app/api/routers/materials.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies import get_current_user, get_session, require_engineer
from app.models.material_profile import MaterialProfile
from app.models.user import User

router = APIRouter(prefix="/materials", tags=["materials"])


class MaterialProfilePayload(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)

    temp_min_z1: Optional[float] = None
    temp_max_z1: Optional[float] = None
    temp_min_z2: Optional[float] = None
    temp_max_z2: Optional[float] = None
    temp_min_z3: Optional[float] = None
    temp_max_z3: Optional[float] = None
    temp_min_z4: Optional[float] = None
    temp_max_z4: Optional[float] = None
    temp_min_z5: Optional[float] = None
    temp_max_z5: Optional[float] = None

    pressure_min: Optional[float] = None
    pressure_max: Optional[float] = None
    speed_min: Optional[float] = None
    speed_max: Optional[float] = None


def _normalize_payload(payload: MaterialProfilePayload) -> Dict[str, Any]:
    data = payload.model_dump()
    data["name"] = data["name"].strip()
    if not data["name"]:
        raise HTTPException(status_code=400, detail="Material name is required")
    return data


def _material_to_dict(material: MaterialProfile) -> Dict[str, Any]:
    return {
        "id": str(material.id),
        "name": material.name,
        "temp_min_z1": material.temp_min_z1,
        "temp_max_z1": material.temp_max_z1,
        "temp_min_z2": material.temp_min_z2,
        "temp_max_z2": material.temp_max_z2,
        "temp_min_z3": material.temp_min_z3,
        "temp_max_z3": material.temp_max_z3,
        "temp_min_z4": material.temp_min_z4,
        "temp_max_z4": material.temp_max_z4,
        "temp_min_z5": material.temp_min_z5,
        "temp_max_z5": material.temp_max_z5,
        "pressure_min": material.pressure_min,
        "pressure_max": material.pressure_max,
        "speed_min": material.speed_min,
        "speed_max": material.speed_max,
        "created_at": material.created_at.isoformat() if material.created_at else None,
        "updated_at": material.updated_at.isoformat() if material.updated_at else None,
    }


@router.get("")
async def list_material_profiles(
    session: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user),
) -> List[Dict[str, Any]]:
    result = await session.execute(
        select(MaterialProfile).order_by(MaterialProfile.created_at.desc())
    )
    materials = result.scalars().all()
    return [_material_to_dict(material) for material in materials]


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_material_profile(
    payload: MaterialProfilePayload,
    session: AsyncSession = Depends(get_session),
    current_user: User = Depends(require_engineer),
) -> Dict[str, Any]:
    material = MaterialProfile(**_normalize_payload(payload))
    session.add(material)
    await session.commit()
    await session.refresh(material)
    return _material_to_dict(material)


@router.put("/{material_id}")
async def update_material_profile(
    material_id: UUID,
    payload: MaterialProfilePayload,
    session: AsyncSession = Depends(get_session),
    current_user: User = Depends(require_engineer),
) -> Dict[str, Any]:
    material = await session.get(MaterialProfile, material_id)
    if material is None:
        raise HTTPException(status_code=404, detail="Material profile not found")

    for field_name, value in _normalize_payload(payload).items():
        setattr(material, field_name, value)

    session.add(material)
    await session.commit()
    await session.refresh(material)
    return _material_to_dict(material)

File: Dev-AI-PM/frontend/src/api/materials.ts
import api from "./index";

export interface MaterialProfilePayload {
    name: string;
    tempMinZ1: number | null;
    tempMaxZ1: number | null;
    tempMinZ2: number | null;
    tempMaxZ2: number | null;
    tempMinZ3: number | null;
    tempMaxZ3: number | null;
    tempMinZ4: number | null;
    tempMaxZ4: number | null;
    tempMinZ5: number | null;
    tempMaxZ5: number | null;
    pressureMin: number | null;
    pressureMax: number | null;
    speedMin: number | null;
    speedMax: number | null;
}

export interface MaterialProfile extends MaterialProfilePayload {
    id: string;
    createdAt: string | null;
    updatedAt: string | null;
}

interface ApiMaterialProfile {
    id: string;
    name: string;
    temp_min_z1: number | null;
    temp_max_z1: number | null;
    temp_min_z2: number | null;
    temp_max_z2: number | null;
    temp_min_z3: number | null;
    temp_max_z3: number | null;
    temp_min_z4: number | null;
    temp_max_z4: number | null;
    temp_min_z5: number | null;
    temp_max_z5: number | null;
    pressure_min: number | null;
    pressure_max: number | null;
    speed_min: number | null;
    speed_max: number | null;
    created_at: string | null;
    updated_at: string | null;
}

const toApiPayload = (payload: MaterialProfilePayload) => ({
    name: payload.name,
    temp_min_z1: payload.tempMinZ1,
    temp_max_z1: payload.tempMaxZ1,
    temp_min_z2: payload.tempMinZ2,
    temp_max_z2: payload.tempMaxZ2,
    temp_min_z3: payload.tempMinZ3,
    temp_max_z3: payload.tempMaxZ3,
    temp_min_z4: payload.tempMinZ4,
    temp_max_z4: payload.tempMaxZ4,
    temp_min_z5: payload.tempMinZ5,
    temp_max_z5: payload.tempMaxZ5,
    pressure_min: payload.pressureMin,
    pressure_max: payload.pressureMax,
    speed_min: payload.speedMin,
    speed_max: payload.speedMax,
});

const fromApiMaterial = (material: ApiMaterialProfile): MaterialProfile => ({
    id: material.id,
    name: material.name,
    tempMinZ1: material.temp_min_z1,
    tempMaxZ1: material.temp_max_z1,
    tempMinZ2: material.temp_min_z2,
    tempMaxZ2: material.temp_max_z2,
    tempMinZ3: material.temp_min_z3,
    tempMaxZ3: material.temp_max_z3,
    tempMinZ4: material.temp_min_z4,
    tempMaxZ4: material.temp_max_z4,
    tempMinZ5: material.temp_min_z5,
    tempMaxZ5: material.temp_max_z5,
    pressureMin: material.pressure_min,
    pressureMax: material.pressure_max,
    speedMin: material.speed_min,
    speedMax: material.speed_max,
    createdAt: material.created_at,
    updatedAt: material.updated_at,
});

export const materialsApi = {
    list: async (): Promise<MaterialProfile[]> => {
        const { data } = await api.get<ApiMaterialProfile[] | { items?: ApiMaterialProfile[]; data?: ApiMaterialProfile[] }>("/materials");
        const items = Array.isArray(data)
            ? data
            : Array.isArray(data?.items)
                ? data.items
                : Array.isArray(data?.data)
                    ? data.data
                    : [];

        return items.map(fromApiMaterial);
    },

    create: async (payload: MaterialProfilePayload): Promise<MaterialProfile> => {
        const { data } = await api.post<ApiMaterialProfile>("/materials", toApiPayload(payload));
        return fromApiMaterial(data);
    },

    update: async (materialId: string, payload: MaterialProfilePayload): Promise<MaterialProfile> => {
        const { data } = await api.put<ApiMaterialProfile>(`/materials/${materialId}`, toApiPayload(payload));
        return fromApiMaterial(data);
    },
};
